### ISSUE TITLE
Allow underscore in test names (suggestion on how to do it)

### ISSUE THREAD

**Issue by cassioneri** [google/googletest#3907]
**Does the feature exist in the most recent commit?**

No. See https://github.com/google/googletest/blob/main/googletest/include/gtest/internal/gtest-internal.h#L1527

**Why do we need this feature?**

The following approach to naming tests is very reasonable:
```
TEST(MyClass, Default_constructor_zero_initialises_all_members)
```
Indeed, the discussion about using `_` in test names comes and goes. See [this post](https://groups.google.com/g/googletestframework/c/N5A07bgEvp4) and  issue #3488 and, of course, the [doc](https://google.github.io/googletest/primer.html#simple-tests) and [FAQ](https://google.github.io/googletest/faq.html#why-should-test-suite-names-and-test-names-not-contain-underscore).

I do understand why `_` is disallowed and that using camel case is a reasonable alternative. However, IMHO, forcing a specific style on clients is a bit too intrusive.

**Describe the proposal**

`googletest` should work harder to avoid conflicts and use other separators that are less likely to be used than `_`. Of course this doesn't eliminate the potential duplication issue. However, by using an ugly string:
```
#define GTEST_TEST_CLASS_NAME_(test_suite_name, test_name) \
  GTEST_SUITE_NAME ## test_suite_name ## GTEST_TEST_NAME ## test_name
```
problems will only occur only if the client uses `GTEST_TEST_NAME` in their test or suit names; or if these names are invalid C++ identifiers (e.g., containing `__` double underscore). 

Ruling out invalid C++ identifiers is already a restriction that forbids say `+` as it does for `__`.

Instead of ruling out `_` altogether, it would be enough to state that (except for comments and strings) the usage of `GTEST` should be avoided. In essence, this rule already exists (possibly undocumented), otherwise client code could interfere with `googletest` internal macros.

**Is the feature specific to an operating system, compiler, or build system version?**

No.

**Objection to the proposal**

When something goes wrong, compilers would report problems with say,
```
GTEST_SUITE_NAMEMyClassGTEST_TEST_NAMEDefault_constructor_zero_initialises_all_members
```
I agree, this is not beautiful but, IMHO, the benefit overweighs this annoyance.

**Comment by derekmauro** [comment]
I agree with the general thoughts of this request, but this may be tricky enough to change without breaking internal Google users that we cannot prioritize it at this time.

**Comment by higher-performance** [comment]
This looks like a duplicate of a newer issue, so I'm closing it. If there is a difference please reopen. (That one has a PR attached, so that's why I picked this to close instead.)
